### PR TITLE
Deneb

### PR DESCRIPTION
**Pull request by moshe-blox**
This PR adds support for Ethereum's Deneb fork.

Most of the changes are due to API refactors in the go-eth2-client package.

### CODE REVIEW

File: go-eth2-client/client.go [hunk 1]
```diff
@@ -83,44 +113,74 @@ func (c *Client) SignedBeaconBlock(ctx context.Context, blockID string) (*spec.V
...
 		return nil, ErrCallNotSupported
 	}
-	return provider.BeaconBlockHeader(ctx, blockID)
+	resp, err := provider.BeaconBlockHeader(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, ErrEmptyResponse
+	}
```
Comment (y0sher) [inline go-eth2-client/client.go:137]: not checking `.Data` here?

File: go-eth2-client/client.go [hunk 2]
```diff
@@ -83,44 +113,74 @@ func (c *Client) SignedBeaconBlock(ctx context.Context, blockID string) (*spec.V
...
 		return nil, ErrCallNotSupported
 	}
-	return provider.SyncCommitteeDuties(ctx, epoch, indices)
+	resp, err := provider.SyncCommitteeDuties(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, ErrEmptyResponse
+	}
```
Comment (y0sher) [inline go-eth2-client/client.go:182]: no need to check `resp.Data` here?

File: go-eth2-client/client.go [hunk 3]
```diff
@@ -131,20 +191,27 @@ func (c *Client) Domain(ctx context.Context, domainType phase0.DomainType, epoch
...
 	}
-	return provider.ValidatorsByPubKey(ctx, stateID, validatorPubKeys)
+	resp, err := provider.Validators(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, ErrEmptyResponse
+	}
+	return resp, nil
```
Comment (y0sher) [inline go-eth2-client/client.go:214]: no need to check `.Data` ? I know its check on the calling function but on some functions here you do check it.

File: go-eth2-client/client.go [hunk 4]
```diff
@@ -163,37 +230,51 @@ func (c *Client) SubmitValidatorRegistrations(ctx context.Context, registrations
...
 	}
-	return provider.BeaconBlockProposal(ctx, slot, randaoReveal, graffiti)
+	resp, err := provider.Proposal(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, ErrEmptyResponse
+	}
+	return resp, nil
```
Comment (y0sher) [inline go-eth2-client/client.go:245]: check `.Data`?

**Review by y0sher — CHANGES_REQUESTED** [review]
some places you check `Data` some not, lets unify

File: go-eth2-client/client.go [hunk 5]
```diff
@@ -83,44 +113,74 @@ func (c *Client) SignedBeaconBlock(ctx context.Context, blockID string) (*spec.V
...
 		return nil, ErrCallNotSupported
 	}
-	return provider.AttesterDuties(ctx, epoch, indices)
+	resp, err := provider.AttesterDuties(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, ErrEmptyResponse
+	}
```
[the remark this answers is not in the document]
Comment (moshe-blox) [reply go-eth2-client/client.go:167]: i tried, however Go's generics are not flexible enough and the wrapper function ends up being too obscure (requires reflection or multiple diff. wrappers), in which case i prefer repetition :-)

File: go-eth2-client/client.go (hunk 1 above)
Comment (moshe-blox) [reply go-eth2-client/client.go:137]: this should preserve the previous functionality, in which the return value (now under `Data`) was not checked.

File: go-eth2-client/client.go (hunk 2 above)
Comment (moshe-blox) [reply go-eth2-client/client.go:182]: this should preserve the previous functionality, in which the return value (now under `Data`) was not checked.

File: go-eth2-client/client.go (hunk 3 above)
Comment (moshe-blox) [reply go-eth2-client/client.go:214]: this should preserve the previous functionality, in which the return value (now under `Data`) was not checked.

File: go-eth2-client/client.go (hunk 4 above)
Comment (moshe-blox) [reply go-eth2-client/client.go:245]: this should preserve the previous functionality, in which the return value (now under `Data`) was not checked.

**Comment by moshe-blox** [comment]
> some places you check `Data` some not, lets unify

replied in comments, basically the idea is that the return value is now under `Data` so i assume that wherever we weren't checking it then we shouldn't have to check `Data` either

**Review by olegshmuelov — APPROVED** [review]

**Review by y0sher — APPROVED** [review]

**Pull request merged**